src/gt4py/next/iterator/type_inference.py

```diff
@@ -157,8 +157,9 @@ def handle_constraint(
                 )
                 for dtype, defined_loc in zip(dtypes, defined_locs)
             ]
-            add_constraint(self.dtypes, Tuple.from_elems(*dtypes))
-            add_constraint(self.defined_locs, Tuple.from_elems(*defined_locs))
+            for dtype, defined_loc in zip(dtypes, defined_locs):
+                add_constraint(self.dtypes, dtype)
+                add_constraint(self.defined_locs, defined_loc)
```

Comment:
TODO make separate PR
Questions:
- Why are these 2 constraints handled separately and not included in the constraint on expanded